src/py/flwr/server/superlink/fleet/grpc_rere/server_interceptor.py

```diff
@@ -147,21 +141,26 @@ def _generic_method_handler(
                     PingRequest,
                 ),
             ):
+                # Verify hmac value
                 hmac_value = base64.urlsafe_b64decode(
                     _get_value_from_tuples(
                         _AUTH_TOKEN_HEADER, context.invocation_metadata()
                     )
                 )
-                client_public_key = bytes_to_public_key(client_public_key_bytes)
-                shared_secret = generate_shared_key(
-                    self.server_private_key,
-                    client_public_key,
-                )
-                verify = verify_hmac(
-                    shared_secret, request.SerializeToString(True), hmac_value
-                )
-                if not verify:
+                public_key = bytes_to_public_key(client_public_key_bytes)
+
+                if not self._verify_hmac(public_key, request, hmac_value):
                     context.abort(grpc.StatusCode.UNAUTHENTICATED, "Access denied")
+
+                # Verify node_id
+                try:
+                    node_id = self.state.get_node_id(client_public_key_bytes)
+                except KeyError:
+                    node_id = -1
+
+                if not self._verify_node_id(node_id, request):
+                    context.abort(grpc.StatusCode.UNAUTHENTICATED, "Access denied")
+
             else:
                 context.abort(grpc.StatusCode.UNAUTHENTICATED, "Access denied")
```

Comment:
Let's unindent the previous section by negating the previous `if` statement and then immediately doing `context.abort`

Reply:
Changed to if not isinstance(request, Request), but we are not reducing number of indents here.

Reply:
I get what you mean, implemented.